src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -483,19 +483,18 @@ public SolutionState AddProject(ProjectInfo projectInfo)
             return this.AddProject(newProject.Id, newProject);
         }
 
-        private ImmutableDictionary<string, ImmutableArray<DocumentId>> CreateLinkedFilesMapWithAddedProject(ProjectState projectState)
+        private ImmutableDictionary<string, ImmutableArray<DocumentId>> CreateFilePathToDocumentIdsMapWithAddedProject(ProjectState projectState)
         {
-            return CreateLinkedFilesMapWithAddedDocuments(projectState, projectState.DocumentIds);
+            var documentIdsAndFilePaths = GetDocumentIdsAndFilePaths(projectState, projectState.DocumentIds, projectState.AdditionalDocumentIds);
+            return CreateFilePathToDocumentIdsMapWithAddedDocuments(documentIdsAndFilePaths);
         }
 
-        private ImmutableDictionary<string, ImmutableArray<DocumentId>> CreateLinkedFilesMapWithAddedDocuments(ProjectState projectState, IEnumerable<DocumentId> documentIds)
+        private ImmutableDictionary<string, ImmutableArray<DocumentId>> CreateFilePathToDocumentIdsMapWithAddedDocuments(IEnumerable<(DocumentId documentId, string filePath)> documentIdsAndFilePaths)
```

Comment:
Any reason this shouldn't just take an `IEnumerable<TextDocumentState>` rather than decomposing into tuples? It'll save you the allocation overhead of the Select methods in GetDocumentIdsAndFilePaths and that method just becomes ProjectState.DocumentStates.Concat(ProjectState.AdditionalDocumentStates) which you can probably re-inline.

Reply:
It also saves you other cases where you're already fetching the state and then decomposing it anyways.